### ISSUE TITLE
backcalls: let's add them

### ISSUE THREAD

**Issue by michaelficarra** [jashkenas/coffeescript#2762]
Copied proposal from https://github.com/jashkenas/coffee-script/issues/2662#issuecomment-13026081 below

---

Taken pretty much from the coffee-of-my-dreams list, the best way to explain my preferred backcall syntax is through example:

<ul><li>

<p>The continuation is captured and tacked on as the final arg by default</p>
<pre lang="coffee">
&lt;- fn a
...
b
</pre>

<p>compiles to</p>

<pre lang="js">
fn(a, function(){
  ...;
  return b;
});
</pre>

</li>
<li>

<p>The left side of the arrow specifies the argument list for the continuation. When only one is given, parens can be omitted.</p>

<pre lang="coffee">
a &lt;- fn b
...
c
</pre>

<p>compiles to</p>

<pre lang="js">
fn(b, function(a){
  ...;
  return c;
});
</pre>

</li>
<li>

<p>And for the rare occasion where the continuation is not the final argument of the function, <del>we need to be able to specify where to put it with a marker like the <code>&lt;&amp;&gt;</code> I use below. I'm open to suggestions for a better marker</del> we can simply use an anonymous function to force the position of the callback.</p>

<pre lang="coffee">
(a, b) &lt;- (_) -&gt; fn c, _, d
...
e
</pre>

<p>compiles to</p>

<pre lang="js">
(function(_){ return fn(c, _, d); })(function(a, b){
  ...;
  return e;
})
</pre>

<p>or, if we detect this case, the simpler:</p>

<pre lang="js">
fn(c, function(a, b){
  ...;
  return e;
}, d);
</pre>

</li></ul>

`this` must be preserved in the continuation by rewriting references as we currently do in bound functions. I don't think we should worry about `arguments` rewriting, but that's your call.

Here's a real-world example @paulmillr gave in #1942:

``` coffee
compile = (files, callback) ->
  async.sortBy files, @sort, (error, sorted) =>
    return @logError error if error?
    async.map sorted, @map, (error, mapped) =>
      return @logError error if error?
      async.reduce mapped, null, @reduce, (error, reduced) =>
        return @logError error if error?
        @write reduced, (error) =>
          return @logError error if error?
          do @log
          callback? @getClassName()
```

and here it is using backcalls:

``` coffee
compile = (files, callback) ->
  (error, sorted) <- async.sortBy files, @sort
  return @logError error if error?
  (error, mapped) <- async.map sorted, @map
  return @logError error if error?
  (error, reduced) <- async.reduce mapped, null, @reduce
  return @logError error if error?
  error <- @write reduced
  return @logError error if error?
  do @log
  callback? @getClassName()
```

**edit:** dropped the callback position indicator, opting for "hugs" style

**Comment by QxQ** [comment]
Just a thought an alternate syntax for this, maybe this would be easier for people to understand?

What if we did "->..." meaning "make a function, but the indentation for it is on the same line", so...

    callme param, (arg) ->...
    codeHere()

is equivalent to

    callme param, (arg) ->
        codeHere()

Also, for the problem with the function parameter in a funny place, it could be solved like this:

    setTimeout ->..., 100
    codeHere()

is equivalent to

    setTimeout ->
        codeHere()
    ,100

**Comment by taoeffect** [comment]
I just saw that ES7 will have `async` and `await` keywords. So what does that imply for coffeescript?

Should I open another issue for `async` and `await` support in CS or is this issue sufficient? cc @jashkenas

**Comment by jashkenas** [comment]
Feel free to open another issue that specifically talks about what adding the ES(7) keywords would look like — and how we might compile them.

I don't think it's something that's possible without large-scale runtime support, no?

**Comment by aurium** [comment]
+1 to @QxQ syntax `->...`

**Comment by Baudin999** [comment]
I appologize for the long comment, please forgive me; but; I really think that a backcall operator obfuscates the code, imagine this:

    result1 <- foo a
    result2 <- foo b
    
Now, does this traslate into:

    function foo(a, function (result1) { /* do something */ }
    function bar(b, function (result2) { /* do something */ }

or:

    function foo(a, function(result1) { 
        function bar(b, function(result2) { /* do something */ })
    })
    
with the proposed syntax you do not know whether or not the second function should be a part of the inner-function of the first or if it should run in parallel. 

I would rather wrap my code in a yield statement like so:

    result1 = yield foo a
    result2 = yield foo b
    
this would resolve into the first option. For the second option you would get:

    await result1 = yield foo a
    result2 = yield foo b

This results into much cleaner code. Now resolving a promise becomes almost trivial:

    promise = do -> foo a
    await result = yield promise.done
    # do something with the result
    console.log result

Now for the problem of the order of the parameters and the callback being somewhere in the middle, combined with mutiple parameters within the callback:

    r1, r2 = do -> foo a, yield, b
    # you could rewrite the first example to:
    result1 = foo a, yield
    # but what would be the point?
    
Here you sacrifice readability for distance to code.

**Comment by 00Davo** [comment]
@Baudin999 Backcalls *always* desugar into the second of your proposed syntaxes, i.e.,
```javascript
a <- thingOne
b <- thingTwo
moreCode
// *always* means
thingOne(function(a) {
  thingTwo(function(b) {
    moreCode;
  });
)};
```

This is consistent with the behaviour of backcalls in Coco and LiveScript, Haskell's `do` notation, and the intuitive sense that a backcall is a "magical async `=` operator". If you want parallel operations, you'll still need a "normal" async helper library. For example:
```coffeescript
{items, otherStuff} <- async.parallel items: getItems, otherStuff: getOtherStuff
# or with promises and a 'p' helper as discussed earlier
results <-p Q.all [firstThing(), secondThing(), thirdThing()]
```

Allowing for distinct `yield` and `await` keywords would be more powerful but also vastly complicates the compiled JavaScript - take a look at IcedCoffeeScript, which provides those exact features, for instance.

**Comment by maxtaco** [comment]
FWIW, I'm contemplating switching IcedCoffeeScript over to a generator-based transpilation, which is of course much simpler. [Here's](https://gist.github.com/maxtaco/05a1f68125ee16540e5c) a hand-compiled example of how it can work.

**Comment by GeoffreyBooth** [comment]
Skimming through this, it seems that the consensus was to not implement this feature. As such, I’m closing this ticket. I think at this point if someone wants to re-propose backcalls, please start a new issue or pull request. Any new proposal should be aimed at the `2` branch, and take into account its support for `async`/`await`.
